moshi/src/main/java/com/squareup/moshi/JsonReader.kt

```diff
@@ -0,0 +1,674 @@
+ * }
+ *
+ * public User readUser(JsonReader reader) throws IOException {
+ *   String username = null;
+ *   int followersCount = -1;
+ *
+ *   reader.beginObject();
+ *   while (reader.hasNext()) {
+ *     String name = reader.nextName();
+ *     if (name.equals("name")) {
+ *       username = reader.nextString();
+ *     } else if (name.equals("followers_count")) {
+ *       followersCount = reader.nextInt();
+ *     } else {
+ *       reader.skipValue();
+ *     }
+ *   }
+ *   reader.endObject();
+ *   return new User(username, followersCount);
+ * }
+ * ```
+ *
+ * ## Number Handling
+ *
+ * This reader permits numeric values to be read as strings and string values to be read as numbers.
+ * For example, both elements of the JSON array `[1, "1"]` may be read using either [nextInt] or [nextString].
+ * This behavior is intended to prevent lossy numeric conversions:
+ * double is JavaScript's only numeric type and very large values like `9007199254740993`
+ * cannot be represented exactly on that platform. To minimize precision loss, extremely large
+ * values should be written and read as strings in JSON.
+ *
+ * Each `JsonReader` may be used to read a single JSON stream. Instances of this class are
+ * not thread safe.
+ */
+public sealed class JsonReader : Closeable {
+  // The nesting stack. Using a manual array rather than an ArrayList saves 20%. This stack will
+  // grow itself up to 256 levels of nesting including the top-level document. Deeper nesting is
+  // prone to trigger StackOverflowErrors.
+  protected var stackSize: Int = 0
+  protected var scopes: IntArray
+  protected var pathNames: Array<String?>
+  protected var pathIndices: IntArray
+  /**
+   * Configure this parser to be liberal in what it accepts. By default, this parser is strict and
+   * only accepts JSON as specified by [RFC 7159](http://www.ietf.org/rfc/rfc7159.txt).
+   * Setting the parser to lenient causes it to ignore the following syntax errors:
+   *
+   * - Streams that include multiple top-level values. With strict parsing, each stream must
+   * contain exactly one top-level value.
+   * - Numbers may be [NaNs][Double.isNaN] or [infinities][Double.isInfinite].
+   * - End of line comments starting with `//` or `#` and ending with a newline
+   * character.
+   * - C-style comments starting with `/ *` and ending with `*``/`. Such
+   * comments may not be nested.
+   * - Names that are unquoted or `'single quoted'`.
+   * - Strings that are unquoted or `'single quoted'`.
+   * - Array elements separated by `;` instead of `,`.
+   * - Unnecessary array separators. These are interpreted as if null was the omitted value.
+   * - Names and values separated by `=` or `=>` instead of `:`.
+   * - Name/value pairs separated by `;` instead of `,`.
+   */
+  @get:CheckReturnValue
+  public var isLenient: Boolean = false
+
+  /** True to throw a [JsonDataException] on any attempt to call [skipValue]. */
+  protected var _failOnUnknown: Boolean = false
```

Comment:
Make `failOnUnknown` will lead to api incompatibilities, have better suggestions?